File: src/test/java/toby/command/commands/music/NowDigOnThisCommandTest.java
package toby.command.commands.music;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import toby.command.CommandContext;

import java.util.concurrent.ArrayBlockingQueue;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

class NowDigOnThisCommandTest implements MusicCommandTest {

    NowDigOnThisCommand nowDigOnThisCommand;

    @BeforeEach
    void setUp() {
        setupCommonMusicMocks();
        nowDigOnThisCommand = new NowDigOnThisCommand();
    }

    @AfterEach
    void tearDown() {
    }

    @Test
    void test_nowDigOnThisCommand_withValidArguments() {
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        when(audioPlayer.isPaused()).thenReturn(false);
        when(playerManager.isCurrentlyStoppable()).thenReturn(false);
        OptionMapping linkOptionalMapping = mock(OptionMapping.class);
        OptionMapping volumeOptionalMapping = mock(OptionMapping.class);
        when(event.getOption("link")).thenReturn(linkOptionalMapping);
        when(event.getOption("volume")).thenReturn(volumeOptionalMapping);
        when(linkOptionalMapping.getAsString()).thenReturn("www.testlink.com");
        when(volumeOptionalMapping.getAsInt()).thenReturn(20);
        ArrayBlockingQueue queue = new ArrayBlockingQueue<>(2);
        AudioTrack track2 = mock(AudioTrack.class);
        when(track2.getInfo()).thenReturn(new AudioTrackInfo("Another Title", "Another Author", 1000L, "identifier", true, "uri"));
        when(track2.getDuration()).thenReturn(1000L);
        when(playerManager.getMusicManager(guild)).thenReturn(musicManager);
        queue.add(track);
        queue.add(track2);
        when(trackScheduler.getQueue()).thenReturn(queue);
        when(track.getUserData()).thenReturn(1);

        //Act
        nowDigOnThisCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        verify(playerManager, times(1)).loadAndPlay(eq(event), eq("www.testlink.com"), eq(false), eq(0), eq(0L), eq(20));
    }

    @Test
    void test_nowDigOnThisCommand_withInvalidPermissionsArguments() {
        setUpAudioChannelsWithBotAndMemberInSameChannel();
        CommandContext commandContext = new CommandContext(event);
        when(audioPlayer.isPaused()).thenReturn(false);
        when(playerManager.isCurrentlyStoppable()).thenReturn(false);
        OptionMapping linkOptionalMapping = mock(OptionMapping.class);
        OptionMapping volumeOptionalMapping = mock(OptionMapping.class);
        when(event.getOption("link")).thenReturn(linkOptionalMapping);
        when(event.getOption("volume")).thenReturn(volumeOptionalMapping);
        when(linkOptionalMapping.getAsString()).thenReturn("www.testlink.com");
        when(volumeOptionalMapping.getAsInt()).thenReturn(20);
        ArrayBlockingQueue queue = new ArrayBlockingQueue<>(2);
        AudioTrack track2 = mock(AudioTrack.class);
        when(track2.getInfo()).thenReturn(new AudioTrackInfo("Another Title", "Another Author", 1000L, "identifier", true, "uri"));
        when(track2.getDuration()).thenReturn(1000L);
        when(playerManager.getMusicManager(guild)).thenReturn(musicManager);
        queue.add(track);
        queue.add(track2);
        when(trackScheduler.getQueue()).thenReturn(queue);
        when(track.getUserData()).thenReturn(1);
        when(requestingUserDto.hasDigPermission()).thenReturn(false);

        //Act
        nowDigOnThisCommand.handleMusicCommand(commandContext, playerManager, requestingUserDto, 0);

        verify(interactionHook, times(1)).sendMessage("I'm gonna put some dirt in your eye Effective Name");
        verify(playerManager, times(0)).loadAndPlay(eq(event), eq("www.testlink.com"), eq(false), eq(0), eq(0L), eq(20));
    }
}

File: src/test/kotlin/toby/command/commands/music/NowDigOnThisCommandTest.kt
package toby.command.commands.music

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo
import io.mockk.*
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import toby.command.CommandContext
import toby.command.CommandTest
import toby.command.CommandTest.Companion.event
import toby.command.CommandTest.Companion.guild
import toby.command.commands.music.MusicCommandTest.Companion.mockAudioPlayer
import toby.command.commands.music.MusicCommandTest.Companion.musicManager
import toby.command.commands.music.MusicCommandTest.Companion.playerManager
import toby.command.commands.music.MusicCommandTest.Companion.track
import toby.command.commands.music.MusicCommandTest.Companion.trackScheduler
import toby.command.commands.music.player.NowDigOnThisCommand
import java.util.concurrent.ArrayBlockingQueue

internal class NowDigOnThisCommandTest : MusicCommandTest {
    private lateinit var nowDigOnThisCommand: NowDigOnThisCommand

    @BeforeEach
    fun setUp() {
        setupCommonMusicMocks()
        nowDigOnThisCommand = NowDigOnThisCommand()
    }

    @AfterEach
    fun tearDown() {
        clearAllMocks()
    }

    @Test
    fun test_nowDigOnThisCommand_withValidArguments() {
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        every { mockAudioPlayer.isPaused } returns false
        every { playerManager.isCurrentlyStoppable } returns false
        every { playerManager.getMusicManager(guild) } returns musicManager

        val linkOptionalMapping = mockk<OptionMapping>()
        val volumeOptionalMapping = mockk<OptionMapping>()
        every { event.getOption("link") } returns linkOptionalMapping
        every { event.getOption("volume") } returns volumeOptionalMapping
        every { event.getOption("start") } returns mockk(relaxed = true)
        every { linkOptionalMapping.asString } returns "www.testlink.com"
        every { volumeOptionalMapping.asInt } returns 20
        every { playerManager.loadAndPlay(any(), any(), any(), any(), any(), any(), any()) } just Runs

        val queue: ArrayBlockingQueue<AudioTrack> = ArrayBlockingQueue(2)
        val track2 = mockk<AudioTrack>()
        every { track2.info } returns AudioTrackInfo(
            "Another Title",
            "Another Author",
            1000L,
            "identifier",
            true,
            "uri"
        )
        every { track2.duration } returns 1000L
        queue.add(track)
        queue.add(track2)
        every { trackScheduler.queue } returns queue
        every { track.userData } returns 1

        // Act
        nowDigOnThisCommand.handleMusicCommand(
            commandContext,
            playerManager,
            CommandTest.requestingUserDto,
            0
        )

        verify(exactly = 1) {
            playerManager.loadAndPlay(
                guild,
                event,
                "www.testlink.com",
                false,
                0,
                0L,
                20
            )
        }
    }

    @Test
    fun test_nowDigOnThisCommand_withInvalidPermissionsArguments() {
        setUpAudioChannelsWithBotAndMemberInSameChannel()
        val commandContext = CommandContext(event)
        every { mockAudioPlayer.isPaused } returns false
        every { playerManager.isCurrentlyStoppable } returns false

        val linkOptionalMapping = mockk<OptionMapping>()
        val volumeOptionalMapping = mockk<OptionMapping>()
        every { event.getOption("link") } returns linkOptionalMapping
        every { event.getOption("volume") } returns volumeOptionalMapping
        every { linkOptionalMapping.asString } returns "www.testlink.com"
        every { volumeOptionalMapping.asInt } returns 20

        val queue: ArrayBlockingQueue<AudioTrack> = ArrayBlockingQueue(2)
        val track2 = mockk<AudioTrack>()
        every { track2.info } returns AudioTrackInfo(
            "Another Title",
            "Another Author",
            1000L,
            "identifier",
            true,
            "uri"
        )
        every { track2.duration } returns 1000L
        every { playerManager.getMusicManager(guild) } returns musicManager
        queue.add(track)
        queue.add(track2)
        every { trackScheduler.queue } returns queue
        every { track.userData } returns 1
        every { CommandTest.requestingUserDto.digPermission } returns false

        // Act
        nowDigOnThisCommand.handleMusicCommand(
            commandContext,
            playerManager,
            CommandTest.requestingUserDto,
            0
        )

        verify(exactly = 1) {
            event.hook.sendMessage("I'm gonna put some dirt in your eye Effective Name")
        }
        verify(exactly = 0) {
            playerManager.loadAndPlay(
                guild,
                event,
                "www.testlink.com",
                false,
                0,
                0L,
                20
            )
        }
    }
}